runtime.md

```diff
@@ -1,5 +1,47 @@
 # Runtime and Lifecycle
 
+## State
+
+The runtime state for a container is persisted on disk so that external tools can consume and act on this information.
+The runtime state is stored in a JSON encoded file.
+It is recommended that this file is stored in a temporary filesystem so that it can be removed on a system reboot.
+On Linux based systems the state information should be stored in `/run/oci/containers`.
+The directory structure for a container is `/run/oci/containers/<containerID>/state.json`.
+By providing a default location that container state is stored external applications can find all containers running on a system.
+
+* **version** (string) Version of the OCI specification used when creating the container.
+* **id** (string) ID is the container's ID.
+* **pid** (int) Pid is the ID of the main process within the container.
+* **root** (string) Root is the path to the container's bundle directory.
+
+The ID is provided in the state because hooks will be executed with the state as the payload.
+This allows the hook to perform clean and teardown logic after the runtime destroys it's own state.
+The root directory to the bundle is provided in the state so that consumers can find the container's configuration and rootfs.
+
+*Example*
+
+```json
+{
+    "id": "oci-container",
+    "pid": 4422,
+    "root": "/containers/redis"
+}
+```
+
+Linux systems add some platform specific information to the state.
+
+* **externalFds** Paths to the container's open file descriptors.
+
+*Example Linux*
+
+```json
+{
+    "externalFds": [
+        "/proc/33/fd/1"
```

Comment:
On Sat, Aug 29, 2015 at 02:58:58AM -0700, Andrew Vagin wrote:
> > So the idea is that externalFds[i] tells you where to connect the
> > restoring runtime's FD i+2.
> 
> ... where to connect the restoring runtime's FD [i].

So externalFds is all the file descriptors, but ExtraFiles is
everything 2 and up.

> > That doesn't cleanly handle cases where we have gaps in the file descriptors
> 
> The libcontainer interface doesn't assume gaps in the file
> descriptors
> https://github.com/golang/go/blob/92390e47d81713d13b0d9890bf8b153f142a5b8a/src/os/exec/exec.go#L82

This isn't a libcontainer spec ;).  I can imagine an OCS runtime that
supports sparse FD arrays, and it's trivial to support in the spec, so
why not allow it here?  If it's too much work for libcontainer to
migrate their array to a map (which seems unlikely), just have
libcontainer explain that it adds an additional continuity requirement
to the externalFds keys.

> > Is “pipe:[18132]” the read or write end of that pipe?
> 
> It doesn't matter. We always can reopen it via /proc.

Who's we?  The runtime?  CRIU?  It's probably not going to be the
application ;).

> > Linking to CRIU docs would be fine for the parts of this that are
> > covered in the CRIU docs.
> 
> http://www.criu.org/Inheriting_FDs_on_restore
> http://criu.org/External_UNIX_socket

While I'm personally interested in those docs, I was asking for docs
in the PR itself (“but I'd like externalFds to land with docs…” [1]).
Nobody will ever come back and read this thread after this PR lands,
but I imagine folks will still want to understand the intended
semantics for externalFds and any implementation limitations.

[1]: https://github.com/opencontainers/specs/pull/87#discussion_r38260688